### ISSUE TITLE
Paginated layout removes arguments from liquid shortcode

### ISSUE THREAD

**Issue by mahnunchik** [11ty/eleventy#2154]
**Describe the bug**

Paginated layout removes arguments from liquid shortcode.

**To Reproduce**

Shortcode
```js
  eleventyConfig.addShortcode('test', function (arg1, arg2, arg3) {
    console.log('test:', arg1, arg2, arg3);
    return `${arg1} - ${arg2} - ${arg3}`;
  });
```

Template `test.liquid`:

```liquid
---
layout: main
pagination:
  data: 'testdata'
  size: 1
  addAllPagesToCollections: true
testdata:
  - item1
  - item2
  - item3
  - item4
---

{% test "file1" "file2" "file3" %}
```

Layout `main.liquid`:
```liquid
{% test "main1" "main2" "main3" %}
{{ content }}
```

**Expected behavior**

All arguments passed to `test` shortcode in both cases: template and layout.

**Actual behavior**

Arguments are randomly removed from shortcode:

```bash
test: file1 file2 file3
test: file1 file2 file3
test: file1 file2 file3
test: file1 file2 file3
test: main1 undefined undefined
test: main1 undefined undefined
test: main1 main2 undefined
test: main1 main3 undefined
```

**Environment:**

- Eleventy Version 1.0.0-beta.X and 1.0.0-canary-X

**Additional context**

Maybe other versions of Eleventy and other template engines are affected.

**Comment by mahnunchik** [comment]
First argument is always present. Arguments from second are removed randomly.

**Comment by pdehaan** [comment]
Interesting. I can reproduce w/ `@beta` and `@canary` channel builds, but not 0.12.1/`@latest` (or beta/canary w/ Nunjucks). Possibly a LiquidJS regression between liquidjs@6 and liquidjs@9 (which I think is the version jump between eleventy@0.12 and eleventy@1).
/cc @harttle 

```sh
> 11ty-2154@1.0.0 build /private/tmp/11ty-2154
> eleventy

www/test/index.html: file1 file2 file3
www/test/1/index.html: file1 file2 file3
www/test/2/index.html: file1 file2 file3
www/test/3/index.html: file1 file2 file3
www/test/2/index.html: main1 undefined undefined
www/test/3/index.html: main1 undefined undefined
www/test/index.html: main1 main2 undefined
www/test/1/index.html: main1 main3 undefined
[11ty] Writing www/test/2/index.html from ./src/test.liquid
[11ty] Writing www/test/3/index.html from ./src/test.liquid
[11ty] Writing www/test/index.html from ./src/test.liquid
[11ty] Writing www/test/1/index.html from ./src/test.liquid
[11ty] Wrote 4 files in 0.06 seconds (v1.0.0-canary.49)
```
Pushed sample repo to https://github.com/pdehaan/11ty-2154

**Comment by harttle** [comment]
Hey @pdehaan, I don't think it's related to liquidjs but tonight I had time to look into it anyway and find it quite interesting.

I added some log to [this file](https://github.com/11ty/eleventy/blob/1ea3a90cc49b636a4c838852a9103e9fe6d2135d/src/Engines/Liquid.js#L95-L125):

```javascript
  static async parseArguments(lexer, str, scope, engine) {
    let argArray = [];
    let label = `[${id++}]`;

    if (!lexer) {
      lexer = moo.compile(Liquid.argumentLexerOptions);
    }

    console.log(label, "parsing arguments for string", str)
    if (typeof str === "string") {
      // TODO key=value key2=value
      // TODO JSON?
      lexer.reset(str);
      let arg = lexer.next();
      while (arg) {
        /*{
          type: 'doubleQuoteString',
          value: '"test 2"',
          text: '"test 2"',
          toString: [Function: tokenToString],
          offset: 0,
          lineBreaks: 0,
          line: 1,
          col: 1 }*/
        if (arg.type.indexOf("ignore:") === -1) {
          const item = await engine.evalValue(arg.value, scope);
          console.log(label, "got item:", arg.value, "eval:", item)
          argArray.push(item);
        }
        arg = lexer.next();
      }
    }
    console.log(label, "return array", argArray)
    return argArray;
  }
```

It prints something like:

```
[4] got item: "main1" eval: main1
[5] got item: "main1" eval: main1
[6] got item: "main1" eval: main1
[6] return array [ 'main1' ]
[7] got item: "main1" eval: main1
[7] return array [ 'main1' ]
[4] got item: "main2" eval: main2
[4] return array [ 'main1', 'main2' ]
[5] got item: "main3" eval: main3
[5] return array [ 'main1', 'main3' ]
```

I assume it's related to the lexer, this line to reset based on string value:

https://github.com/11ty/eleventy/blob/1ea3a90cc49b636a4c838852a9103e9fe6d2135d/src/Engines/Liquid.js#L105

But in JavaScript strings are immutable, which means the three passes of `parseArguments()` are reseting each other regardless they're different references to that string. Can be confirmed if I comment out `if (!lexer)` to create a new lexer each pass. I suppose the following result is expected.

```
www/test/index.html: file1 file2 file3
www/test/1/index.html: file1 file2 file3
www/test/2/index.html: file1 file2 file3
www/test/3/index.html: file1 file2 file3
www/test/index.html: main1 main2 main3
www/test/1/index.html: main1 main2 main3
www/test/2/index.html: main1 main2 main3
www/test/3/index.html: main1 main2 main3
[11ty] Writing www/test/index.html from ./src/test.liquid
[11ty] Writing www/test/1/index.html from ./src/test.liquid
[11ty] Writing www/test/2/index.html from ./src/test.liquid
[11ty] Writing www/test/3/index.html from ./src/test.liquid
[11ty] Wrote 4 files in 0.05 seconds (v1.0.0-canary.49)
```

The reason why it's may be related to LiquidJS version could be something changed to async and now concurrency actually happens. Nevertheless, I guess it can be a bug to reset the lexer based on string values. First time dig into this so maybe I'm wrong. @zachleat please let me know if there's something need my help.

**Comment by pdehaan** [comment]
Awesome, thanks for the 👀  @harttle! ❤️

**Comment by thisanimus** [comment]
I'm also noticing this behavior in 1.x eleventy builds when using a template that extends another template.  eg:

If I extend the base layout in `_layouts/posts.liquid`: 
```
---
layout: base.liquid
---
```

I get the same random missing args

**Comment by epelc** [comment]
Are there any plans to apply the fix from @harttle ? We just ran into this and our layouts have a ton of shortcodes.

Would you accept a PR with this change?

**Comment by epelc** [comment]
I just figured out a proper fix for this and #2348 without having to create a new lexer on each call.

Instead of using await on each call to evalValue you have to let it consume the current scope and use a `await Promise.all()` instead. This makes each call actually consume it's current scope at the time of call to `evalValue` otherwise it is randomly using the wrong scope because scope's fields are updated between calls to `parseArguments` since it's a object and objects are by reference in javascript.

I will submit a PR tomorrow but this is the rough code.
Without fix:
<img width="502" alt="image" src="https://user-images.githubusercontent.com/5204642/167071493-008ad990-ffe5-44d5-a119-311b47810e90.png">

With fix:
<img width="473" alt="image" src="https://user-images.githubusercontent.com/5204642/167071435-47ba8ca7-3d09-42e6-a0a1-4a7f5a41afc0.png">

Thanks @harttle for pointing in the right direction.

**Comment by zachleat** [comment]
Looks like we have two different conversations going so I’ll post comments on both 😅

> Fixed by #2367. Thank you @epelc!

> Release versions here: https://github.com/11ty/eleventy/pull/2367#issuecomment-1119872993

**Comment by zachleat** [comment]
And a thank you to @harttle too!!

This reminded me also to set up an 11ty sponsorship of Liquid.js cc https://github.com/11ty/11ty-community/issues/78
